File: frontend/src/utils/errorHandler.ts
/**
 * Centralized error handling utilities for API errors.
 */

export interface APIError {
  error: string;
  message: string;
  type: string;
}

interface ValidationErrorItem {
  loc?: string[];
  msg?: string;
}

interface StructuredErrorDetail {
  message?: string;
  type?: string;
  error?: string;
}

interface AxiosErrorResponse {
  response?: {
    data?: {
      detail?: string | ValidationErrorItem[] | StructuredErrorDetail;
    };
  };
  message?: string;
}

function isAxiosError(error: unknown): error is AxiosErrorResponse {
  return typeof error === 'object' && error !== null && 'response' in error;
}

function isStructuredDetail(detail: unknown): detail is StructuredErrorDetail {
  return (
    typeof detail === 'object' &&
    detail !== null &&
    ('message' in detail || 'type' in detail || 'error' in detail)
  );
}

function isValidationErrorArray(detail: unknown): detail is ValidationErrorItem[] {
  return Array.isArray(detail);
}

export class ErrorHandler {
  /**
   * Extract user-friendly error message from various error formats.
   */
  static extractMessage(error: unknown): string {
    if (isAxiosError(error) && error.response?.data?.detail) {
      const detail = error.response.data.detail;

      if (isStructuredDetail(detail) && detail.message) {
        return detail.message;
      }

      if (isValidationErrorArray(detail)) {
        const messages = detail.map((err) => {
          const loc = err.loc?.join(' → ') || 'unknown';
          const msg = err.msg || 'Invalid value';
          return `${loc}: ${msg}`;
        });
        return messages.join('\n');
      }

      if (typeof detail === 'string') {
        return detail;
      }
    }

    if (error instanceof Error) {
      return error.message;
    }

    if (typeof error === 'string') {
      return error;
    }

    return 'An unexpected error occurred. Please try again.';
  }

  /**
   * Extract error type from error response.
   */
  static extractType(error: unknown): string {
    if (isAxiosError(error)) {
      const detail = error.response?.data?.detail;
      if (isStructuredDetail(detail) && detail.type) {
        return detail.type;
      }
    }
    return 'unknown_error';
  }

  /**
   * Check if error is a validation error.
   */
  static isValidationError(error: unknown): boolean {
    const type = ErrorHandler.extractType(error);
    return type === 'validation_error' || type === 'value_error';
  }

  /**
   * Check if error is a server error.
   */
  static isServerError(error: unknown): boolean {
    const type = ErrorHandler.extractType(error);
    return type === 'server_error';
  }

  /**
   * Format error for display in UI.
   */
  static formatForDisplay(error: unknown): {
    title: string;
    message: string;
    type: string;
  } {
    const message = ErrorHandler.extractMessage(error);
    const type = ErrorHandler.extractType(error);

    let title = 'Error';

    if (ErrorHandler.isValidationError(error)) {
      title = 'Validation Error';
    } else if (ErrorHandler.isServerError(error)) {
      title = 'Server Error';
    }

    if (isAxiosError(error)) {
      const detail = error.response?.data?.detail;
      if (isStructuredDetail(detail) && detail.error) {
        title = detail.error;
      }
    }

    return { title, message, type };
  }
}

File: backend/app/core/error_handler.py
"""Centralized error handling utilities."""

from pydantic import ValidationError
from fastapi import HTTPException


class ValidationErrorHandler:
    """Handles validation errors with user-friendly messages."""

    @staticmethod
    def format_pydantic_error(error: ValidationError) -> str:
        """Convert Pydantic validation errors to user-friendly messages."""
        messages = []

        for err in error.errors():
            error_type = err["type"]
            msg = err.get("msg", "")

            # If error message already includes parameter name (from our custom validator)
            # use it directly
            if error_type == "value_error" and "Parameter '" in msg:
                messages.append(msg)
                continue

            # Otherwise, build the message
            field_path = " → ".join(str(loc) for loc in err["loc"])
            value = err.get("input")

            if error_type == "decimal_parsing":
                messages.append(
                    f"Parameter '{field_path}': Expected a numeric value, "
                    f"got '{value}' instead. Please enter a valid number (e.g., 0.5, 100, 3.14)."
                )
            elif error_type == "decimal_type":
                messages.append(
                    f"Parameter '{field_path}': Expected a numeric value, "
                    f"got {type(value).__name__} instead. Please enter a valid number."
                )
            elif error_type == "missing":
                messages.append(f"Parameter '{field_path}': This field is required.")
            elif error_type == "string_type":
                messages.append(
                    f"Parameter '{field_path}': Invalid value '{value}'. {msg}"
                )
            else:
                # Generic fallback - use message from validator if available
                if msg:
                    messages.append(msg)
                else:
                    messages.append(f"Parameter '{field_path}': Invalid value")

        return "\n".join(messages)

    @staticmethod
    def to_http_exception(
        error: Exception, status_code: int = 400
    ) -> HTTPException:
        """Convert various errors to HTTPException with structured response."""

        if isinstance(error, ValidationError):
            message = ValidationErrorHandler.format_pydantic_error(error)
            return HTTPException(
                status_code=status_code,
                detail={
                    "error": "Validation Error",
                    "message": message,
                    "type": "validation_error",
                },
            )

        elif isinstance(error, ValueError):
            return HTTPException(
                status_code=status_code,
                detail={
                    "error": "Invalid Input",
                    "message": str(error),
                    "type": "value_error",
                },
            )

        else:
            # Don't expose internal errors to users
            return HTTPException(
                status_code=500,
                detail={
                    "error": "Internal Server Error",
                    "message": "An unexpected error occurred. Please try again or contact support.",
                    "type": "server_error",
                },
            )
